kong/runloop/handler.lua

```diff
@@ -78,13 +78,26 @@ do
     return false
   end
 
+
+  local function load_service_from_db(service_pk)
+    return kong.db.services:select(service_pk)
+  end
+
   local function get_service_for_route(db, route)
     local service_pk = route.service
     if not service_pk then
       return nil
     end
 
-    local service, err = db.services:select(service_pk)
+    local service, err
+    if kong.cache then -- the kong.cache module is sometimes missing during tests
+      local cache_key = kong.db.services:cache_key(service_pk.id)
+      service, err = kong.cache:get(cache_key, CACHE_ROUTER_OPTS,
+                                    load_service_from_db, service_pk)
+    else
+      service, err = load_service_from_db(service_pk)
+    end
+
     if not service then
       return nil, "could not find service for route (" .. route.id .. "): " ..
```

Comment:
Sometimes as we've seen, `service` will be `nil` not because of an error, but because it is not found (eventual consistency between the Routes/Services relationship). We should probably distinguish `if err then ... elseif not service then ... end` and produce a different log in each branch (avoiding the concatenation of a `nil` value which supersede the underlying error. Something to keep in mind as we work on other router-related PRs.